src/librustc_typeck/check/mod.rs

```diff
@@ -3898,28 +3916,56 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         fn_decl: &hir::FnDecl,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
-        can_suggest: bool,
+        can_suggest: FnDeclType,
     ) -> bool {
         // Only suggest changing the return type for methods that
         // haven't set a return type at all (and aren't `fn main()` or an impl).
         match (&fn_decl.output, found.is_suggestable(), can_suggest, expected.is_unit()) {
-            (&hir::FunctionRetTy::DefaultReturn(span), true, true, true) => {
+            (&hir::FunctionRetTy::DefaultReturn(span), true, FnDeclType::Item, true) => {
                 err.span_suggestion(
                     span,
                     "try adding a return type",
                     format!("-> {} ", self.resolve_type_vars_with_obligations(found)),
                     Applicability::MachineApplicable);
                 true
             }
-            (&hir::FunctionRetTy::DefaultReturn(span), false, true, true) => {
+            (&hir::FunctionRetTy::DefaultReturn(span), false, FnDeclType::Item, true) => {
                 err.span_label(span, "possibly return type missing here?");
                 true
             }
-            (&hir::FunctionRetTy::DefaultReturn(span), _, false, true) => {
-                // `fn main()` must return `()`, do not suggest changing return type
+            (&hir::FunctionRetTy::DefaultReturn(span), _, FnDeclType::ImplItem, true) => {
+                // do not suggest changing return type for trait impls
                 err.span_label(span, "expected `()` because of default return type");
                 true
             }
+            (&hir::FunctionRetTy::DefaultReturn(span), _, FnDeclType::MainItem, true) => {
+                // in main(), we can only return types that implement `std::process::Termination`
+                let ret_ty = self.resolve_type_vars_with_obligations(found);
+                let is_term = if let Some(term_id) = self.infcx.tcx.lang_items().termination() {
+                    self.infcx.predicate_must_hold_modulo_regions(
+                        &self.infcx.tcx.predicate_for_trait_def(
+                            self.param_env,
+                            traits::ObligationCause::dummy(),
+                            term_id,
+                            0,
+                            ret_ty,
+                            &[]
+                        ))
+                } else {
+                    false
+                };
+                if is_term {
+                    err.span_suggestion(
+                        span,
+                        "try adding a return type",
+                        format!("-> {} ", ret_ty),
```

Comment:
I think the whitespace in the code suggestion might be subtly wrong. I believe this should be `" -> {} ", but without the `stderr` test output I can't say for sure.

Reply:
I just added a UI test, but it appears that the suggestion doesn't appear. Any idea why?

Reply:
I'm not sure. Could you add some `debug` statements to verify the contents of `fn_decl.output`, `found.is_suggestable()`, `can_suggest`, `expected` and `ret_ty`?

Reply:
Sure thing.